cmd/state-tray/internal/menu/projects.go

```diff
@@ -30,44 +31,71 @@ func (u *LocalProjectsUpdater) Update(projects []*graph.Project) {
 		item.remove()
 	}
 
-	u.items = make([]*localProjectsMenuItem, 0)
-	for _, project := range projects {
-		u.addLocalProject(project)
+	u.items = make([]*localProjectsMenuItem, len(projects))
+	for i, project := range projects {
+		item := newLocalProjectMenuItem(project, u.menuItem)
+		u.items[i] = item
+		item.startEventLoop()
 	}
 }
 
-func (u *LocalProjectsUpdater) addLocalProject(project *graph.Project) {
-	ctx, cancel := context.WithCancel(context.Background())
-
-	mProject := u.menuItem.AddSubMenuItem(project.Namespace, "")
-	u.items = append(u.items, &localProjectsMenuItem{mProject, project, cancel})
-
-	if len(project.Locations) == 1 {
-		go waitForClick(ctx, mProject, project.Namespace, project.Locations[0])
-		return
+func newLocalProjectMenuItem(project *graph.Project, menuItem *systray.MenuItem) *localProjectsMenuItem {
+	done := make(chan struct{})
+	mProject := menuItem.AddSubMenuItem(project.Namespace, "")
+	item := &localProjectsMenuItem{
+		menuItem:  mProject,
+		namespace: project.Namespace,
+		done:      done,
+		children:  []*localProjectsMenuItem{},
 	}
 
-	for _, location := range project.Locations {
+	for i, location := range project.Locations {
+		if i == 0 && i == len(project.Locations)-1 {
+			item.location = location
+			break
+		}
+
 		mLocation := mProject.AddSubMenuItem(location, "")
-		go waitForClick(ctx, mLocation, project.Namespace, location)
+		item.addChild(mLocation, location)
 	}
+
+	return item
 }
 
-func (u *localProjectsMenuItem) remove() {
-	u.cancel()
-	u.menuItem.Hide()
+func (i *localProjectsMenuItem) addChild(menuItem *systray.MenuItem, location string) {
+	i.children = append(i.children, &localProjectsMenuItem{
```

Comment:
Use your constructor rather than manually creating a `localProjectsMenuItem{}` again.

If your constructor can't service the use case you can create a wrapper constructor. Key point being that the type is only constructed in one single function unless there's some strong reason to do otherwise.

This will likely force you to refactor your constructor, but that's not a bad thing.